### ISSUE TITLE
Erroneous trait bound not satisfied error

### ISSUE THREAD

**Issue by changhe3** [intellij-rust/intellij-rust#7459]
## Environment

* **IntelliJ Rust plugin version:**
* 0.4.150.3968-211
* **Rust toolchain version:**
* 1.53
* **IDE name and version:**
* CLion 211.7442.42
* **Operating system:**
* Windows 10

## Problem description
This code produces a false positive type error:
```rust
pub trait IterResult: Iterator<Item = Result<Self::Ok, Self::Error>> + Sized {
    type Ok;
    type Error;
}

impl<I, T, E> IterResult for I
where
    I: Iterator<Item = Result<T, E>>,
{
    type Ok = T;
    type Error = E;
}
```

Error: Line 6
```
the trait bound `I: Iterator<Item=Result<Ok, Error>>` is not satisfied [E0277] 
the trait `Iterator<Item=Result<Ok, Error>>` is not implemented for `I`
```

**Comment by Seeker14491** [comment]
I ran into a similar error. I reduced it to this:

![2021-07-22_18-49](https://user-images.githubusercontent.com/6490497/126723110-8166262a-309a-4c26-add1-0151ebaf1511.png)

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d2c740b84197b10536ed024af6cba327)

**Comment by drewcassidy** [comment]
Seeing the same thing implementing the `Zero` trait on a class where `Add<Self>` is implemented as part of a blanket impl
